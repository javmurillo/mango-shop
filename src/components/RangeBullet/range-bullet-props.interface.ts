export interface RangeBulletProps {
  offset: string;
  factor: number;
  handleRef:
    | string
    | ((instance: HTMLDivElement | null) => void)
    | React.RefObject<HTMLDivElement>
    | null
    | undefined;
  handleMove: Function;
  step: number;
}
