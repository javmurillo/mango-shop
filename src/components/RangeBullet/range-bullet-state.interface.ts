export interface RangeBulletState {
  hovered: boolean;
  focused: boolean;
  active: boolean;
  activeCount?: number;
}
