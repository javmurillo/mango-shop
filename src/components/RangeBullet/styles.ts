export interface BulletStyles {
  handle: Object;
  focusedHandle: Object;
  hoveredHandle: Object;
  activeHandle: Object;
}

export const bulletStyles: BulletStyles = {
  handle: {
    position: 'absolute',
    display: 'inline-block',
    cursor: 'pointer',
    height: 30,
    width: 30,
    backgroundColor: 'black',
    border: '1px solid #052350',
    borderRadius: '50%',
  },
  focusedHandle: {
    border: '2px solid 052350',
  },
  hoveredHandle: {
    backgroundColor: 'white',
    border: '2px solid #052350',
    boxShadow: '0px 0px 5px 0px gray',
    height: '38px',
    width: '38px',
  },
  activeHandle: {
    backgroundColor: 'white',
    border: '2px solid #052350',
    cursor: 'grab',
    boxShadow: 'inset 0px 0px 5px 0px gray',
    height: '35px',
    width: '35px',
  },
};
