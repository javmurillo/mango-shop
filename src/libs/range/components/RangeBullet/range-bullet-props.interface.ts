export interface RangeBulletProps {
  offset: string;
  factor: number;
  handleRef?:
    | string
    | ((instance: HTMLDivElement | null) => void)
    | React.RefObject<HTMLDivElement>
    | null
    | undefined;
  handleMove?: (increase: number) => void;
  step: {
    left: number;
    right: number;
  };
  ariaLabel?: string;
}
