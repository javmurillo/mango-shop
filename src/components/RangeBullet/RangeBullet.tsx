import { Component } from 'react';
import { RangeBulletProps } from './range-bullet-props.interface';
import { RangeBulletState } from './range-bullet-state.interface';
import { bulletStyles } from './styles';
import { CSSProperties } from 'react';

export default class RangeBullet extends Component<
  RangeBulletProps,
  RangeBulletState
> {
  public state: RangeBulletState = {
    hovered: false,
    focused: false,
    active: false,
  };

  private style: CSSProperties;
  private lastPos: number | undefined;
  private currentPos: number | undefined;

  constructor(props: RangeBulletProps) {
    super(props);
    this.style = {
      ...bulletStyles.handle,
      left: this.props.offset,
    };
  }

  componentDidMount(): void {
    document.addEventListener('mousemove', this.onBulletMouseMove);
    document.addEventListener('mouseup', this.onBulletMouseUp);
  }

  componentWillReceiveProps(props: RangeBulletProps): void {
    this.style = this.getBulletStyle(this.state, props);
  }

  onMouseEnterBullet = (): void => {
    this.style = this.getBulletStyle(
      { ...this.state, hovered: true },
      this.props
    );
    this.setState({
      hovered: true,
    });
  };

  onMouseLeaveBullet = (): void => {
    this.style = this.getBulletStyle(
      { ...this.state, hovered: false },
      this.props
    );
    this.setState({
      hovered: false,
    });
  };

  onMouseDownBullet = (
    event: React.MouseEvent<HTMLDivElement, MouseEvent>
  ): void => {
    this.setBulletToStartMoving(event, event.pageX);
  };

  onBulletMouseMove = (event: MouseEvent): void => {
    if (this.state.active) {
      this.move(event, event.clientX);
    }
  };

  /*
   *  If the range bullet is active, just stop it where the event mouse up was fired.
   */
  onBulletMouseUp = (event: MouseEvent): void => {
    if (this.state.active) {
      this.move(event, event.pageX);
      this.moveEnd();
    }
  };

  setBulletToStartMoving = (
    event: MouseEvent | React.MouseEvent<HTMLDivElement, MouseEvent>,
    position: number
  ): void => {
    event.preventDefault();
    event.stopPropagation();
    this.style = this.getBulletStyle(
      { ...this.state, active: true },
      this.props
    );
    this.currentPos = position;
    this.lastPos = position;
    this.setState({
      active: true,
    });
  };

  move = (event: MouseEvent, position: number): void => {
    event.preventDefault();
    event.stopPropagation();
    const { factor, step, handleMove } = this.props;
    const direction = position - this.lastPos!;
    const distance = position - this.currentPos!;

    const increment = direction > 0 ? 1 : -1;
    if (direction * distance > (factor || 1) * step) {
      handleMove(increment);
      this.currentPos! += factor * step * increment;
    }
    this.lastPos = position;
  };

  moveEnd = (): void => {
    this.style = this.getBulletStyle(
      { ...this.state, active: false },
      this.props
    );
    this.setState({
      active: false,
    });
  };

  getBulletStyle(state: RangeBulletState, props: RangeBulletProps) {
    const { hovered, focused, active } = state;
    const { offset } = props;
    const calcHoverStyle = hovered ? bulletStyles.hoveredHandle : undefined;
    const calcFocusStyle = focused ? bulletStyles.focusedHandle : undefined;
    const calcActiveStyle = active ? bulletStyles.activeHandle : undefined;

    return {
      ...bulletStyles.handle,
      ...calcHoverStyle,
      ...calcFocusStyle,
      ...calcActiveStyle,
      left: offset,
    };
  }

  render(): JSX.Element {
    return (
      <div
        ref={this.props.handleRef}
        style={this.style}
        onMouseEnter={this.onMouseEnterBullet}
        onMouseLeave={this.onMouseLeaveBullet}
        onMouseDown={this.onMouseDownBullet}
      ></div>
    );
  }
}
