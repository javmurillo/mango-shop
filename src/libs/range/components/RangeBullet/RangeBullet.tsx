import { Component, CSSProperties } from 'react';
import { RangeBulletProps } from './range-bullet-props.interface';
import { RangeBulletState } from './range-bullet-state.interface';
import { bulletStyles } from './styles';

/**
 * RangeBullet class
 */
export default class RangeBullet extends Component<
  RangeBulletProps,
  RangeBulletState
> {
  // RangeBullet state
  public state: RangeBulletState = {
    hovered: false,
    focused: false,
    active: false,
  };

  private style: CSSProperties; // CSS Style for this component.
  private lastPos: number | undefined; // Last bullet position.
  private currentPos: number | undefined; // Current bullet position.

  constructor(props: RangeBulletProps) {
    super(props);
    // Style initialization
    this.style = {
      ...bulletStyles.handle,
      left: this.props.offset,
    };
  }

  componentDidMount(): void {
    document.addEventListener('mousemove', this.onBulletMouseMove);
    document.addEventListener('mouseup', this.onBulletMouseUp);
  }

  /**
   * @event onMouseEnter Hover style activated.
   */
  onMouseEnterBullet = (): void => {
    this.style = this.getBulletStyle(
      { ...this.state, hovered: true },
      this.props
    );
    this.setState({
      hovered: true,
    });
  };

  /**
   * @event onMouseLeave Hover style deactivated.
   */
  onMouseLeaveBullet = (): void => {
    this.style = this.getBulletStyle(
      { ...this.state, hovered: false },
      this.props
    );
    this.setState({
      hovered: false,
    });
  };

  /**
   * Sets the bullet ready to start moving.
   * @event onMouseDown Hover style deactivated.
   */
  onMouseDownBullet = (
    event: React.MouseEvent<HTMLDivElement, MouseEvent>
  ): void => {
    this.setBulletToStartMoving(event, event.pageX);
  };

  /**
   * Moves the bullet if active.
   * @event mousemove
   * @param event MouseEvent
   */
  onBulletMouseMove = (event: MouseEvent): void => {
    if (this.state.active) {
      this.move(event, event.clientX);
    }
  };

  /**
   * If the range bullet is active, just stop it where the event mouse up was fired.
   * @param event MouseEvent
   */
  onBulletMouseUp = (event: MouseEvent): void => {
    if (this.state.active) {
      this.move(event, event.pageX);
      this.moveEnd();
    }
  };

  /**
   * Updates the state setting [active] to true, and setting the [currentPos] and [lastPos] global properties.
   * @param event MouseEvent | React.MouseEvent<HTMLDivElement, MouseEvent>
   * @param position number
   */
  setBulletToStartMoving = (
    event: MouseEvent | React.MouseEvent<HTMLDivElement, MouseEvent>,
    position: number
  ): void => {
    event.preventDefault();
    event.stopPropagation();
    this.style = this.getBulletStyle(
      { ...this.state, active: true },
      this.props
    );
    this.currentPos = position;
    this.lastPos = position;
    this.setState({
      active: true,
    });
  };

  /**
   * Moves the bullet.
   * @param event MouseEvent
   * @param position Current bullet position
   */
  move = (event: MouseEvent, position: number): void => {
    event.preventDefault();
    event.stopPropagation();
    const { factor, step, handleMove } = this.props;
    const direction = position - this.lastPos!;
    const distance = position - this.currentPos!;
    const increment = direction > 0 ? 1 : -1;
    const directionStep = increment === 1 ? step.right : step.left;
    if (direction * distance > (factor || 1) * directionStep) {
      if (handleMove) {
        handleMove(increment);
      }
      this.currentPos! += factor * directionStep * increment;
    }
    this.lastPos = position;
  };

  /**
   * Updates the state setting [active] to false, finishing the movement.
   */
  moveEnd = (): void => {
    this.style = this.getBulletStyle(
      { ...this.state, active: false },
      this.props
    );
    this.setState({
      active: false,
    });
  };

  /**
   * Returns the new style given the current state and properties.
   * @param state RangeBulletState
   * @param props RangeBulletProps
   */
  getBulletStyle(state: RangeBulletState, props: RangeBulletProps): Object {
    const { hovered, focused, active } = state;
    const { offset } = props;
    const calcHoverStyle = hovered ? bulletStyles.hoveredHandle : undefined;
    const calcFocusStyle = focused ? bulletStyles.focusedHandle : undefined;
    const calcActiveStyle = active ? bulletStyles.activeHandle : undefined;

    return {
      ...bulletStyles.handle,
      ...calcHoverStyle,
      ...calcFocusStyle,
      ...calcActiveStyle,
      left: offset,
      zIndex: props.zIndex,
    };
  }

  render(): JSX.Element {
    this.style = this.getBulletStyle(this.state, this.props);

    return (
      <div
        aria-label={this.props.ariaLabel}
        ref={this.props.handleRef}
        style={this.style}
        onMouseEnter={this.onMouseEnterBullet}
        onMouseLeave={this.onMouseLeaveBullet}
        onMouseDown={this.onMouseDownBullet}
      ></div>
    );
  }
}
