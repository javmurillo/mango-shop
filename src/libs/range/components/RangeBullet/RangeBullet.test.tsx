import { cleanup, render } from '@testing-library/react';
import TestRenderer from 'react-test-renderer';
import RangeBullet from './RangeBullet';

afterEach(cleanup);

const setup = () => {
  const utils = render(
    <RangeBullet
      factor={1}
      offset="100"
      step={{ left: 0, right: 50 }}
      ariaLabel="range bullet"
    />
  );
  const rangeBullet = utils.getByLabelText('range bullet');
  return {
    rangeBullet,
    ...utils,
  };
};

describe('<RangeBullet/> tests', () => {
  test('renders correctly', () => {
    const { rangeBullet } = setup();

    expect(rangeBullet).toBeInTheDocument();
  });

  test('state initialized correctly', () => {
    const { rangeBullet } = setup();

    expect(rangeBullet).toBeInTheDocument();
  });

  test('shoud match snapshot', () => {
    const testRendererJson = TestRenderer.create(
      <RangeBullet factor={1} offset="100" step={{ left: 0, right: 50 }} />
    ).toJSON();
    expect(testRendererJson).toMatchSnapshot();
  });
});
