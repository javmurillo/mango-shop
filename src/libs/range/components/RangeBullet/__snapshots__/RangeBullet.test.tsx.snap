// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RangeBullet/> tests shoud match snapshot 1`] = `
<div
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "backgroundColor": "black",
      "border": "1px solid #052350",
      "borderRadius": "50%",
      "cursor": "grab",
      "display": "inline-block",
      "height": 30,
      "left": "100",
      "position": "absolute",
      "width": 30,
    }
  }
/>
`;
